lib/irb/type_completion/scope.rb

```diff
@@ -0,0 +1,408 @@
+# frozen_string_literal: true
+
+require 'set'
+require_relative 'types'
+module IRB
+  module TypeCompletion
+    class BaseScope
+      BREAK_RESULT = '%break'
+      NEXT_RESULT = '%next'
+      RETURN_RESULT = '%return'
+      PATTERNMATCH_BREAK = '%match'
+
+      attr_reader :module_nesting, :self_object
+
+      def initialize(binding, self_object, local_variables)
+        @binding = binding
+        @self_object = self_object
```

Comment:
This is just a reminder: `self_object` is `BasicObject`, many methods here will break. But IRB doesn't really work inside a `BasicObject` anyway (other things would break first), we can probably get away with it for now.